driver/base/DatagramDevice.zig

```diff
@@ -0,0 +1,180 @@
+//!
+//! An abstract datagram orientied device with runtime dispatch.
+//!
+//! Datagram devices behave similar to an SPI or Ethernet device where
+//! packets with an ahead-of-time known length can be transferred in a
+//! single transaction.
+//!
+
+const std = @import("std");
+
+const DatagramDevice = @This();
+
+const BaseError = error{ IoError, Timeout };
+
+pub const ConnectError = BaseError || error{DeviceBusy};
+
+/// Establishes a connection to the device (like activating a chip-select lane or similar).
+/// NOTE: Call `.disconnect()` when the usage of the device is done to release it.
+pub fn connect(dd: DatagramDevice) ConnectError!void {
+    if (dd.vtable.connectFn) |connectFn| {
+        return connectFn(dd.object);
+    }
+}
+
+/// Releases a device from the connection.
+pub fn disconnect(dd: DatagramDevice) void {
+    if (dd.vtable.disconnectFn) |disconnectFn| {
+        return disconnectFn(dd.object);
+    }
+}
+
+pub const WriteError = BaseError || error{ Unsupported, NotConnected };
+
+/// Writes a single `datagram` to the device.
+pub fn write(dd: DatagramDevice, datagram: []const u8) WriteError!void {
+    if (dd.vtable.writeFn) |writeFn| {
+        return writeFn(dd.object, datagram);
+    } else {
+        return error.Unsupported;
+    }
+}
+
+pub const ReadError = BaseError || error{ Unsupported, NotConnected };
+
+/// Reads a single `datagram` from the device.
+pub fn read(dd: DatagramDevice, datagram: []u8) ReadError!void {
+    if (dd.vtable.readFn) |readFn| {
+        return readFn(dd.object, datagram);
+    } else {
+        return error.Unsupported;
+    }
+}
+
+object: ?*anyopaque,
```

Comment:
please put struct members at the top of the file, it's difficult to find them in this sea of text